[Kernel] Add parseStructType to the JsonHandler for parsing Delta schemas

allisonport-db opened the pull request:
<!--
Thanks for sending a pull request!  Here are some tips for you:
  1. If this is your first time, please read our contributor guidelines: https://github.com/delta-io/delta/blob/master/CONTRIBUTING.md
  2. If the PR is unfinished, add '[WIP]' in your PR title, e.g., '[WIP] Your PR title ...'.
  3. Be sure to keep the PR description updated to reflect all changes.
  4. Please write your PR title to summarize what this PR proposes.
  5. If possible, provide a concise example to reproduce the issue for a faster review.
  6. If applicable, include the corresponding issue number in the PR title and link it in the body.
-->

#### Which Delta project/connector is this regarding?
<!--
Please add the component selected below to the beginning of the pull request title
For example: [Spark] Title of my pull request
-->

- [ ] Spark
- [ ] Standalone
- [ ] Flink
- [X] Kernel
- [ ] Other (fill in here)

## Description

The current version of Kernel has a temporary data type [MixedType](https://github.com/delta-io/delta/blob/master/kernel/kernel-api/src/main/java/io/delta/kernel/internal/types/MixedDataType.java) to help with parsing Delta schemas. This PR updates Kernel to use the solution decided in this doc: https://docs.google.com/document/d/1UOlw498twX0OME1T4e66lV9daZz7OjY4YCw0WxK2Jzs/edit?usp=sharing.

- Adds method `parseStructType` to the `JsonHandler` interface
- Removes `MixedDataType` and `TableSchemaSerDe` as they are unneeded using the new API
- Implements `parseStructType` in the `DefaultJsonHandler`

Resolves delta-io/delta#2277

## How was this patch tested?

Adds unit tests.

vkorukanti commented on kernel/kernel-api/src/main/java/io/delta/kernel/client/JsonHandler.java:
```diff
@@ -52,6 +52,17 @@ public interface JsonHandler
      */
     ColumnarBatch parseJson(ColumnVector jsonStringVector, StructType outputSchema);
 
+    /**
+     * Parse the given {@code schemaString} according to the Delta Protocol
+     * <a href="https://github.com/delta-io/delta/blob/master/PROTOCOL.md#primitive-types">
+     *    serialization rules </a>.
+     *
+     * @param schemaString the JSON formatted schema string to parse
+     * @return the parsed {@link StructType}
```
```suggestion
    /**
     * Deserialize the Delta table schema from {@code schemaString} according to the Delta Protocol
     * <a href="https://github.com/delta-io/delta/blob/master/PROTOCOL.md#primitive-types">
     *    schema serialization rules</a>.
     *
     * @param schemaString the JSON formatted schema string to parse
     * @return the parsed {@link StructType}
```

scottsand-db replied:
@vkorukanti - if you want the text to say `Deserialize the Delta table schema ...` why don't we rename this method to `deserializeTableSchema`?

vkorukanti commented on kernel/kernel-api/src/main/java/io/delta/kernel/client/JsonHandler.java:
```diff
@@ -52,6 +52,17 @@ public interface JsonHandler
+     * @return the parsed {@link StructType}
+     */
+    // TODO solidify method name
+    StructType parseStructType(String schemaString);
```
`deserializeStructType`? no strong opinions. 

`serializedStructType` name can be the counterpart for the other way.

scottsand-db replied:
```java
deserializeTableSchema(String schemaJson)
```

allisonport-db replied:
I think we don't want to mention "table schema". But we could do either `deserializeStructType` or `deserializeDeltaSchema`

scottsand-db replied:
```
deserializeStructType(String structTypeJson)
```

SGTM

vkorukanti commented on kernel/kernel-api/src/main/java/io/delta/kernel/types/FieldMetadata.java:
```diff
@@ -0,0 +1,203 @@
+/**
+ * The metadata for a given {@link StructField}.
+ */
+public final class FieldMetadata {
```
What extra functionality do we get with this class? Wondering it is worth the extra class.

vkorukanti commented on kernel/kernel-api/src/test/java/io/delta/kernel/internal/types/JsonHandlerTestImpl.java:
```diff
@@ -50,6 +50,11 @@ public CloseableIterator<FileReadContext> contextualizeFileReads(
+    }
+
     @Override
     public ColumnarBatch parseJson(ColumnVector jsonStringVector, StructType outputSchema) {
```
Do we need this class anymore?

allisonport-db replied:
Good catch we can finally remove it

vkorukanti commented on kernel/kernel-defaults/src/main/java/io/delta/kernel/defaults/internal/DataTypeParser.java:
```diff
@@ -0,0 +1,295 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.delta.kernel.defaults.internal;
```
nit: types.internal is a better place?

allisonport-db replied:
Assuming you meant `internal.types`

vkorukanti commented on kernel/kernel-defaults/src/main/java/io/delta/kernel/defaults/internal/DataTypeParser.java:
```diff
@@ -0,0 +1,295 @@
+            return (StructType) parsedType;
+        } else {
+            throw new IllegalArgumentException(
+                String.format("Could not parse %s as a valid StructType", json));
```
given `json` could be multi-line? in that is the case print it at the end of the message.

scottsand-db replied:
`Could not parse the following JSON as a valid StructType:\n%s`

vkorukanti commented on kernel/kernel-defaults/src/main/java/io/delta/kernel/defaults/internal/DataTypeParser.java:
```diff
@@ -0,0 +1,295 @@
+                return nameToType(json.textValue());
+            case OBJECT:
+                // complex types (array, map, or struct are stored as JSON objects)
+                String type = getStringField(json, "type");
```
consider if lowercasing is needed here.

allisonport-db replied:
I'm thinking we don't _need_ accomodate for this case as it wouldn't be a valid schema but we still could. What do you think?

vkorukanti commented on kernel/kernel-defaults/src/main/java/io/delta/kernel/defaults/internal/DataTypeParser.java:
```diff
@@ -0,0 +1,295 @@
+                }
+            default:
+                throw new IllegalArgumentException(
+                    String.format("Could not parse %s as a valid Delta data type", json));
```
same comment as above

vkorukanti commented on kernel/kernel-defaults/src/test/scala/io/delta/kernel/defaults/client/DataTypeParserSuite.scala:
```diff
@@ -0,0 +1,275 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.delta.kernel.defaults.client
```
package in internal.

allisonport-db replied:
moved to `io.delta.kernel.defaults.internal.types`

vkorukanti commented on kernel/kernel-defaults/src/test/scala/io/delta/kernel/defaults/client/DefaultJsonHandlerSuite.scala:
```diff
@@ -0,0 +1,110 @@
+    assert(e.getMessage.contains("Could not parse JSON"))
+  }
+
+  // TODO we use toJson to serialize our physical and logical schemas in ScanStateRow, we should
```
Looks like you have migrated the tests from the deleted suite `TestTableSchemaSerDe` here which uses `DataType.toJson`. Do we need more tests? If not can we remove the TODO here?

allisonport-db replied:
I think the two tests aren't really enough. But that code is all in kernelAPI still and will be replaced later so not completely sure what we should do here.

vkorukanti replied:
Got it. Could you please create issues for this TODO and the other TODO about FieldMetadata.

allisonport-db replied:
https://github.com/delta-io/delta/issues/2313 https://github.com/delta-io/delta/issues/2315

vkorukanti reviewed (changes requested):
Overall lgtm. 

One question about FieldMetadata. Seems like it is adding a lot of code. Wondering do we need to copy the Delta-Spark API in Kernel? 

Also, check what type of exception should the new API throw in case of invalid schema. Currently, the implementation throws `IllegalArgumentException`, but the API doesn't declare any. Should we create a new exception class `InvalidSerializedSchemaException` that can custom print the error message. It can have the overall given schema and the specific part that is wrong.

scottsand-db commented on kernel/kernel-api/src/main/java/io/delta/kernel/types/StructField.java:
```diff
@@ -59,13 +56,23 @@ public class StructField {
+    private final FieldMetadata metadata;
+
+    public StructField(
+        String name,
+        DataType dataType,
+        boolean nullable) {
+        this.name = name;
```
nit: indentation

scottsand-db commented on kernel/kernel-api/src/main/java/io/delta/kernel/types/StructField.java:
```diff
@@ -59,13 +56,23 @@ public class StructField {
         DataType dataType,
         boolean nullable,
-        Map<String, String> metadata) {
+        FieldMetadata metadata) {
```
nit: fix indentation while you're at it

scottsand-db replied:
> Adds FieldMetadata class to store struct field metadata. We cannot use an ordinary Map here since field metadata stores values of different types.

Can this be done separately? In a refactor PR _before_ this PR?

scottsand-db reviewed (changes requested):
My biggest question here is why the `FieldMetadata` is part of this PR. Can you please explain this in the PR description?

scottsand-db commented:
> I'm happy to factor out separate PR that just adds `FieldMetadata`.

Yup, let's do this please. It'll make this PR much simpler and cohesive.

allisonport-db commented:
@scottsand-db @vkorukanti I've removed `FieldMetadata` from this PR and updated the `DataTypeParser` to maintain the current behavior. I will add the `FieldMetadata` class, parsing for it, and tests in a follow-up PR.

scottsand-db commented on kernel/kernel-api/src/main/java/io/delta/kernel/client/JsonHandler.java:
```diff
@@ -52,6 +52,16 @@ public interface JsonHandler
     ColumnarBatch parseJson(ColumnVector jsonStringVector, StructType outputSchema);
 
+    /**
+     * Deserialize the Delta schema from {@code schemaString} according to the Delta Protocol
```
nit: `{@code schemaString` should probably be `{@code structTypeJson}`, eh

scottsand-db commented on kernel/kernel-defaults/src/main/java/io/delta/kernel/defaults/internal/types/DataTypeParser.java:
```diff
@@ -0,0 +1,234 @@
+        return metadata;
+    }
+
+    private static String FIXED_DECIMAL = "decimal\\(\\s*(\\d+)\\s*,\\s*(\\-?\\d+)\\s*\\)";
```
what should _really_ be static here isn't just the string, but the `Pattern.compile(FIXED_DECIMAL)`. That's the expensive part that we don't want to recompute every `nameToType` invocation

allisonport-db replied:
Fixed

scottsand-db commented on kernel/kernel-defaults/src/main/java/io/delta/kernel/defaults/internal/types/DataTypeParser.java:
```diff
@@ -0,0 +1,234 @@
+     * }
+     * </pre>
+     */
+    protected static DataType parseDataType(JsonNode json) {
```
Why `protected static`? Who is extending this?

https://stackoverflow.com/questions/24289070/why-we-should-not-use-protected-static-in-java

allisonport-db replied:
for testing, made it package private

allisonport-db commented on kernel/kernel-defaults/src/main/java/io/delta/kernel/defaults/internal/types/DataTypeParser.java:
```diff
@@ -0,0 +1,234 @@
+import static io.delta.kernel.internal.util.Preconditions.checkArgument;
+
+/**
+ * Parses Delta data types based on the
```
Updated

vkorukanti commented on kernel/kernel-defaults/src/main/java/io/delta/kernel/defaults/internal/types/DataTypeParser.java:
```diff
@@ -0,0 +1,234 @@
+                    case "array":
+                        return parseArrayType(json);
+                    case "map":
+                        return parseMapType(json);
```
nit: add a comment for not having the default case here
